toolchain/check/inst.cpp

```diff
@@ -193,6 +222,32 @@ auto EvalOrAddInst(Context& context, SemIR::LocIdAndInst loc_id_and_inst)
   }
 }
 
+auto AddInstWithoutEval(Context& context, SemIR::LocIdAndInst loc_id_and_inst)
+    -> SemIR::ConstantId {
+  CARBON_CHECK(!loc_id_and_inst.inst.kind().has_cleanup());
+  switch (loc_id_and_inst.inst.kind().constant_needs_inst_id()) {
+    case SemIR::InstConstantNeedsInstIdKind::No: {
+      return AddImportedConstant(context, loc_id_and_inst.inst);
+    }
+
+    case SemIR::InstConstantNeedsInstIdKind::DuringEvaluation: {
+      // Evaluation temporarily needs an InstId. Add one for now.
+      auto inst_id = AddInstInNoBlockWithOptionalEval(context, loc_id_and_inst,
+                                                      WithoutEval);
+      // TODO: Consider removing `inst_id` from `insts` if it's still the most
+      // recently added instruction.
+      return context.constant_values().Get(inst_id);
+    }
+
+    case SemIR::InstConstantNeedsInstIdKind::Permanent: {
+      // Evaluation needs a permanent InstId. Add the instruction.
+      auto inst_id =
+          AddInstWithOptionalEval(context, loc_id_and_inst, WithoutEval);
+      return context.constant_values().Get(inst_id);
+    }
+  }
+}
```

Comment:
The semantics of this function seem inconsistent:
* In the first case here, we add it to the constants store.
* In the second case, we add it to no block as an orphan.
* In the third case, we add it to the current block.

It seems like we should probably be providing only a single behavior here (perhaps the first one?), regardless of what evaluation needs, because we're not invoking evaluation.

If we don't pick the third of the above semantics, this function seems to have the wrong name. (The third one is a form of `AddInst`, the second is a form of `AddInstInNoBlock`; for the first thing I might consider something more like `AddReducedConstantWithoutEval`?)